### PR TITLE
SNOW-878073 Add MaxRetryCount as configurable parameter

### PR DESCRIPTION
**Pull request by sfc-gh-dheyman-1**
### Description
Added MaxRetryCount as configurable parameter. It specifies how many times requests can be retried.

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [ ] Created tests which fail without the change (if possible)
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: dsn_test.go [hunk 1]
```diff
@@ -339,6 +339,22 @@ func TestParseDSN(t *testing.T) {
 			},
 			ocspMode: ocspModeFailOpen,
 		},
+		{
+			dsn: "u:p@a?database=d&maxRetryCount=20",
```
Comment (sfc-gh-pfus) [inline dsn_test.go:343]: Can we verify that the default value is correctly set, if not configured?

File: retry_test.go [hunk 2]
```diff
@@ -353,7 +353,37 @@ func TestRetryQueryFail(t *testing.T) {
 	}
 	_, err = newRetryHTTP(context.Background(),
 		client,
-		emptyRequest, urlPtr, make(map[string]string), 15*time.Second, defaultTimeProvider, nil).doPost().setBody([]byte{0}).execute()
+		emptyRequest, urlPtr, make(map[string]string), 15*time.Second, 100, defaultTimeProvider, nil).doPost().setBody([]byte{0}).execute()
+	if err == nil {
```
Comment (sfc-gh-pfus) [inline retry_test.go:357]: We have an assertion library for such checks :)

File: retry_test.go [hunk 3]
```diff
@@ -353,7 +353,37 @@ func TestRetryQueryFail(t *testing.T) {
...
+	values, err = url.ParseQuery(urlPtr.RawQuery)
+	if err != nil {
+		t.Fatalf("failed to parse the URL: %v", err)
+	}
+	retry, err := strconv.Atoi(values.Get(retryCountKey))
+	if err != nil {
+		t.Fatalf("failed to get retry counter: %v", err)
+	}
+	if retry < 2 {
+		t.Fatalf("not enough retry counter: %v", retry)
```
Comment (sfc-gh-pfus) [inline retry_test.go:370]: Is it more like `not enough retires attempted`?

File: dsn_test.go (hunk 1 above)
Comment (sfc-gh-dheyman-1) [reply dsn_test.go:344]: Added a check in other test.

File: retry_test.go (hunk 2 above)
Comment (sfc-gh-dheyman-1) [reply retry_test.go]: Changed nil checks in this file to use `assertNilf/assertNotNilf`

File: retry_test.go (hunk 3 above)
Comment (sfc-gh-dheyman-1) [reply retry_test.go]: Changed the name to be more applicable.

**Review by sfc-gh-pfus — APPROVED** [review]

**Pull request merged**